text/alpha-modules.md

```diff
@@ -0,0 +1,141 @@
+> APIs `A` and `B` and they have backwards incompatible revisions (`A1`, `A2`,
+> `A3`, ...). Furthermore, let's say that if we leave old copies of the API in
+> place in new releases of a library (but `@deprecated`), we're going to write
+> that as `(A1) + A2` (in this case, `A1` is deprecated but still in the
+> library, and `A2` is also in there²).
+
+> Let's say these are the releases of `cdk-experiments-change`:
+>
+> | cdk-experiments-change@0.1.0 | cdk-experiments-change@0.2.0 | cdk-experiments-change@0.3.0 |
+> | ---------------------------- | ---------------------------- | ---------------------------- |
+> | `A1 + B1`                    | `A2 + B1`                    | `A2 + B2`                    |
+
+> First we change `A`, then we change `B`.
+
+> Now, assume that we have an application or library, doesn't really matter
+> which, that's happily consuming `A1`. They are on library version `0.1.0`, and
+> all is swell.
+
+> Now `B2` just got a feature that they want to use while staying on `A1` for
+> the `A` API but... that's impossible! There's no package version that supplies
+> `A1 + B2`.
+
+> And this is _potentially_ fine, you can just update your code to migrate from
+> `A1` -> `A2` instead as well, right? No big deal.
+
+> It becomes more hairy once we start mixing in libraries that you may not have
+> control over. Different scenario:
+
+> Maybe you are an application that uses a library that hard-locked you to
+> `0.1.0` because it uses `B1` (and you can't know whether `0.2.0` will still
+> have `B1` or not, so you have to minor-version-lock). But now the app wants to
+> use `A2`. We're in the scenario that there's there _is_ a version that
+> supplies `A2 + B1`, and yet you're **still not able to take it** because your
+> library dependency locked you into a version of the CDK lib.
+
+```
+app —[ uses L1 ]⟶ lib@1.0.0 —[ uses B1 ]⟶ cdk-experiments-change@0.1.0
+app —[ uses B2 ]⟶ cdk-experiments-change@0.2.0                    **** CONFLICT ****
+```
+
+> Bottom line:
+
+> **Building ecosystems around in-place broken APIs get _worse_ if we tie those
+> APIs together in releases.**
+
+**Should unstable modules be allowed to depends on other unstable modules?**
+
+Two observations:
+
+1. In both options, stable modules can not depend on unstable modules.
+2. Due to the above, in order to graduate a module and add it to `aws-cdk-lib`,
+   we will have to graduate all of its dependencies.
+
+We know from our current dependency tree, that unstable modules depending on
+other unstable modules is a valid use case. From the 35 unstable modules we
+currently have, ~3 of them depends on other unstable modules.
+
+**If we release unstable modules separately can they depend on each other?**
+
+Yes. Unstable module can depend on other unstable modules using
+`peerDependencies`.
+
+The main concern raised in the PR was that doing so will result in "the same
+dependency hell monocdk was meant to solve". As commented on the PR, this is not
```

Comment:
This also brings up again the fact that those will be exactly like any 3rd party module in the ecosystem so if we have an issue it applies to the entire ecosystem